src/avro_binary_decoder.erl

```diff
@@ -29,79 +29,185 @@
+        fun(__Type__, __SubNameOrId__, Data, DecodeFun) -> DecodeFun(Data) end).
+
+-type lkup_fun() :: fun((string()) -> avro_type()).
+
 %%%_* APIs =====================================================================
 
+%% @doc decode/4 equivalent with default hook fun.
+-spec decode(iodata(), string() | avro_type(), lkup_fun()) -> term().
+decode(IoData, Type, Lkup) when is_function(Lkup, 1) ->
+  decode(IoData, Type, Lkup, ?DEFAULT_HOOK).
+
 %% @doc Decode bytes into unwrapped avro value, assuming the input bytes
 %% matches the given schema without tailing bytes.
 %% @end
 -spec decode(iodata(), string() | avro_type(),
-             fun((string()) -> avro_type())) -> term().
-decode(IoData, Type, LookupFun) ->
+             lkup_fun(), hook_fun()) -> term().
+decode(IoData, Type, Lkup, Hook) when is_function(Lkup, 1) ->
   %% return decoded value as raw erlang term directly
-  {Value, <<>>} = do_decode(IoData, Type, LookupFun),
+  {Value, <<>>} = do_decode(IoData, Type, Lkup, Hook),
   Value.
 
-%% @doc Decode the header of a byte stream,
-%% return unwrapped value and tail bytes in a tuple.
-%% @end
+%% @doc decode_stream/4 equivalent with default hook fun.
 -spec decode_stream(iodata(), string() | avro_type(),
                     fun((string()) -> avro_type())) -> term().
-decode_stream(IoData, Type, LookupFun) ->
-  do_decode(IoData, Type, LookupFun).
+decode_stream(IoData, Type, Lkup) when is_function(Lkup, 1) ->
+  decode_stream(IoData, Type, Lkup, ?DEFAULT_HOOK).
+
+%% @doc Decode the header of a byte stream, return unwrapped value and tail
+%% bytes in a tuple.
+%% @end
+decode_stream(IoData, Type, Lkup, Hook) when is_function(Lkup, 1) ->
+  do_decode(IoData, Type, Lkup, Hook).
+
+%% @doc Return a function to be used as the decode hook.
+%% The hook prints the type tree with indentation, and the leaf values.
+%% @end
+-spec pretty_print_debug_hook() -> hook_fun().
+pretty_print_debug_hook() ->
+  fun(T, SubInfo, Data, DecodeFun) ->
+    Name = avro:get_type_fullname(T),
+    Indentation =
+      case get(avro_binary_decoder_pp_indentation) of
+        undefined -> 0;
+        Indentati -> Indentati
+      end,
+    IndentationStr = lists:duplicate(Indentation * 2, $\s),
+    ToPrint =
+      [ IndentationStr
+      , Name
+      , case SubInfo of
+          []                   -> ": ";
+          I when is_integer(I) -> [$., integer_to_list(I), "\n"];
+          S                    -> [$., S, "\n"]
+        end
+      ],
+    io:format("~s", [ToPrint]),
+    _ = put(avro_binary_decoder_pp_indentation, Indentation + 1),
+    {Result, Tail} = DecodeFun(Data),
+    %% print empty array and empty map
+    case SubInfo =/= [] andalso Result =:= [] of
+      true  -> io:format("~s  []\n", [IndentationStr]);
+      false -> ok
+    end,
+    %% print the value if it's a leaf in the type tree
+    _ = SubInfo =:= [] andalso io:format("~1000000p\n", [Result]),
+    _ = put(avro_binary_decoder_pp_indentation, Indentation),
```

Comment:
+1 before going deep into the stack, and set back after popped back up to the current depth in the stack.